internal/manifest/oci_index.go

```diff
@@ -76,6 +78,26 @@ func (index *OCI1IndexPublic) UpdateInstances(updates []ListUpdate) error {
 	return index.editInstances(editInstances)
 }
 
+func addCompressionAnnotations(compressionAlgorithms []compression.Algorithm, annotationsMap map[string]string) {
+	foundGzip := false
+	for _, algo := range compressionAlgorithms {
+		switch algo.Name() {
+		case compression.ZstdAlgorithmName:
+			annotationsMap[OCI1InstanceAnnotationCompressionZSTD] = OCI1InstanceAnnotationCompressionZSTDValue
+		case compression.GzipAlgorithmName:
+			foundGzip = true
+		default:
+			continue
+		}
+	}
+
+	// In case if user is reverting to default gzip
+	// we must remove other inbuilt annotatations
+	if foundGzip {
```

Comment:
I can’t see how this follows. AFAICS the zstd annotation should be used iff a zstd layer is present, and presence or absence of gzip layers has nothing to do with it.

(Consider a future in which there is some other third format.)

Reply:
@mtrmac Yes I had some confusion about how to implement deletion since there is now way to determine if user wants to remove `zstd` annotation. Any hints about how to add deletion here ?

Reply:
The idea is that the caller provides a _complete and precise_ list of compression algorithms. (That doesn’t currently exist.)

Then this code can always add or remove the zstd annotation automatically. The caller is a generic OCI-unaware thing that doesn’t know a zstd annotation exists; the caller only provides the requires list of algorithms.

---

… and that doesn’t work for the existing `UpdateInstances`. Oops.

---

A point to note is that this only works if the list of algorithms is specified. I.e., right now, this PR would cause the delete code to keep deleting the annotation.

So, either split this further (per https://github.com/containers/image/pull/1992#discussion_r1228373189 ):
- revert this PR to only add an annotation
- implement correctly setting `*CompressionAlgorithms` in the caller
- and then re-add the part of this PR that deletes the zstd annotation.

Or, do all of the three in _this_ PR, i.e. extend it to correctly set `*CompressionAlghorithms` in c/image/copy.

Reply:
> revert this PR to only add an annotation

SGTM I think doing this later on will be easier once we have some working.